syntax = "proto3";

package protos.raft;

service Raft {
  rpc AppendEntries (AppendEntriesArgs) returns (AppendEntriesReply) {}
  rpc RequestVote (RequestVoteArgs) returns (RequestVoteReply) {}
}

message LogEntry {
  uint64 log_index = 1;
  uint64 log_term = 2;

  string log_name = 3;
  bytes data = 4;
  bool command_valid = 5;
}

message AppendEntriesArgs {
  uint64 term = 1;
  int32 leader_id = 2;
  uint64 prev_log_index = 3;

  uint64 prev_log_term = 4;
  repeated LogEntry entries = 5;

  uint64 leader_commit_index = 6;
}

message AppendEntriesReply {
  uint64 term = 1;
  bool success = 2;

  uint64 update_index = 3;

  uint64 conflict_index = 4;
  uint32 conflict_term = 5;
}

message RequestVoteArgs {
  uint64 term = 1;
  int32 candidate_id = 2;
  uint64 last_log_index = 3;
  uint64 last_log_term = 4;
}

message RequestVoteReply {
  uint64 term = 1;
  bool vote_granted = 2;
}

message Snapshot {
  uint64 last_included_index = 1;
  uint64 last_included_term = 2;
  bytes snapshot = 3;
}